metadata/metadata.go

```diff
@@ -309,30 +338,62 @@ func (c *Client) WriteGuestAttributes(ctx context.Context, key, value string) er
 	return err
 }
 
-// GetKey gets a specific metadata key.
-func (c *Client) GetKey(ctx context.Context, key string) (string, error) {
-	req, err := http.NewRequest("GET", c.metadataURL+key, nil)
+func (c *Client) do(ctx context.Context, cfg requestConfig) (string, error) {
+	var (
+		urlTokens []string
+		finalURL  string
+	)
+
+	if cfg.hang {
+		urlTokens = append(urlTokens, "wait_for_change=true")
+		urlTokens = append(urlTokens, "last_etag="+c.etag)
+	}
+
+	if cfg.timeout > 0 {
+		urlTokens = append(urlTokens, fmt.Sprintf("timeout_sec=%d", cfg.timeout))
+	}
+
+	if cfg.recursive {
+		urlTokens = append(urlTokens, "recursive=true")
+	}
+
+	if cfg.jsonOutput {
+		urlTokens = append(urlTokens, "alt=json")
+	}
+
+	finalURL = cfg.baseURL
+
+	if len(urlTokens) > 0 {
+		finalURL = fmt.Sprintf("%s/?%s", finalURL, strings.Join(urlTokens, "&"))
+	}
+
+	req, err := http.NewRequestWithContext(ctx, cfg.method, finalURL, nil)
 	if err != nil {
 		return "", err
 	}
 
 	req.Header.Add("Metadata-Flavor", "Google")
-	req = req.WithContext(ctx)
-
-	res, err := c.httpClient.Do(req)
-	if err == nil {
-		// TODO: Expand and propagate this error handling to the other metadata operations.
-		errorMsgFmt := "error connecting to metadata server, status code: %d"
-		switch res.StatusCode {
-		case 404, 412:
-			return "", fmt.Errorf(errorMsgFmt, res.StatusCode)
-		}
-	} else if err != nil {
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
 		return "", fmt.Errorf("error connecting to metadata server: %+v", err)
 	}
 
-	defer res.Body.Close()
-	md, err := ioutil.ReadAll(res.Body)
+	statusCodeMsg := "error connecting to metadata server, status code: %d"
+	switch resp.StatusCode {
+	case 404, 412:
+		return "", fmt.Errorf(statusCodeMsg, resp.StatusCode)
```

Comment:
should this be returning error for all non-2xx status codes?

Reply:
it should, but I'll have to review all the "special codes" for now we are only handling this two, but others will follow.